app/controllers/users/two_factor_authentication/setup_controller.rb

```diff
@@ -0,0 +1,79 @@
+class Users::TwoFactorAuthentication::SetupController < ApplicationController
+  # Skips 2FA so User can set up the totp via QR code
+  skip_before_action :handle_two_factor_authentication
+  # Skips 2FA setup confirmation callback in ApplicationController.
+  skip_before_action :confirm_two_factor_setup
+
+  def show
+    if params[:method] == "email"
+      render_email_setup
+    elsif params[:method] == "app"
+      render_app_setup
+    else
+      render :show
+    end
+  end
+
+  def update
+    if params[:method] == "email"
+      save_two_factor_method "email"
+      send_email
+      redirect_to "/users/two_factor_authentication/email"
+      return
+    end
+
+    @otp_secret_key = params[:otp_secret_key]
+    if current_user.authenticate_totp(params[:code], otp_secret_key: @otp_secret_key)
+      current_user.otp_secret_key = @otp_secret_key
+      current_user.save(validate: false)
+      save_two_factor_method "app"
+      flash[:notice] = "Two factor authentication setup successful"
+      disable_2fa_checks_for_session
+    else
+      flash[:alert] = "Six digit code is not valid"
+      # TODO: Replace with route helper once fully migrate to new 2FA routes.
+      redirect_back(fallback_location: "/users/two_factor_authentication/setup")
+    end
+  end
+
+  def qr_code_uri
+    provisioning_uri = current_user.provisioning_uri(
```

Comment:
what about the `provisioning_uri` from https://github.com/Houdini/two_factor_authentication#customisation-and-usage

Reply:
This is from the existing 2FA code. If we think this can be improved, we can certainly mark it as tech debt, but I'd rather not change it in this PR.